qt: Force set nPruneSize in QSettings after the intro dialog

hebasto opened the pull request:
On master (5622d8f3156a293e61d0964c33d4b21d8c9fd5e0), having `QSettings` set already
```
$ grep nPruneSize ~/.config/Bitcoin/Bitcoin-Qt-testnet.conf
nPruneSize=6
```

enabling prune option in the intro dialog

```
$ ./src/qt/bitcoin-qt -choosedatadir -testnet
```
![DeepinScreenshot_select-area_20191208120425](https://user-images.githubusercontent.com/32963518/70388183-eed68580-19b6-11ea-9aa1-f9ad9aaa68a6.png)

has no effect:
```
$ grep Prune ~/.bitcoin/testnet3/debug.log
2019-12-08T10:04:41Z Prune configured to target 5722 MiB on disk for block and undo files.
```

---

With this PR:
```
$ grep Prune ~/.bitcoin/testnet3/debug.log 
2019-12-08T10:20:35Z Prune configured to target 1907 MiB on disk for block and undo files.
```

This PR has been split of #17453 (the first two commits) as it fixes an orthogonal bug.

Refs:
- https://github.com/bitcoin/bitcoin/pull/17453#discussion_r345424240
- https://github.com/bitcoin/bitcoin/pull/17453#discussion_r350960201

hebasto committed:
util: Replace magics with DEFAULT_PRUNE_TARGET_GB

This commit does not change behavior.

hebasto committed:
qt: Force set nPruneSize in QSettings after intro

If QSettings is set already, it is required to force set nPruneSize
after the intro dialog.

Sjors commented:
Tested ACK 6928c50fa3104ce1b77c5d12928d49584fa536d5

In macOS the settings are in `~/Library/Preferences/org.bitcoin.Bitcoin-Qt-regtest.plist` which can be viewed with Xcode.

jonasschnelli commented:
utACK 6928c50fa3104ce1b77c5d12928d49584fa536d5

Sjors commented:
@promag can you do a review?

hebasto commented:
@fanquake @MarcoFalke 
Would you mind reviewing this PR?

promag commented:
I have tested that this change resets to 2GB regardless of what's in the settings.

However I'm not sure this is correct, actually I think current behavior is correct - only problem is that the intro dialog doesn't show what's in the settings. In your case the intro should have the checkbox checked and it should say "6 GB (prune)".

I don't think running with `-choosedatadir` should make it ignore existing settings - for that we have `-resetguisettings`?

Sjors commented:
> I don't think running with -choosedatadir should make it ignore existing settings - for that we have -resetguisettings?

I don't mind either way; the plan seems to be to store settings in the datadir, e.g. using rw_config #11082.

promag commented:
> the plan seems to be to store settings in the datadir

In that case the current prune setting would still be available.

hebasto commented:
@ryanofsky Would you mind reviewing this PR w.r.t. future of settings handling?

Sjors commented:
> > the plan seems to be to store settings in the datadir
> 
> In that case the current prune setting would still be available.

It doesn't make sense to me to keep the same prune setting for a different data directory; it's probably on a different disk.

ryanofsky reviewed (approved):
Code review ACK 6928c50fa3104ce1b77c5d12928d49584fa536d5. I think the change makes sense and does fix a surprising behavior. Thanks for splitting this from the other PR and describing the problem so clearly here!

I think the alternate approach suggested by promag https://github.com/bitcoin/bitcoin/pull/17696#issuecomment-570261615 to just fix the displayed size and not change the behavior would also address the bug, and maybe be less surprising for someone not expecting `-choosedatadir` to affect the prune size. But as Sjors points out there are multiple reasons to expect that changing the datadir would affect the size, so I think that as long as the displayed size is correct, using either the current value or default value seems ok.

hebasto commented on src/qt/bitcoin.cpp:
```diff
@@ -282,7 +282,7 @@ void BitcoinApplication::parameterSetup()
 void BitcoinApplication::SetPrune(bool prune, bool force) {
-     optionsModel->SetPrune(prune, force);
+    optionsModel->SetPruneTargetGB(prune ? DEFAULT_PRUNE_TARGET_GB : 0, force);
 }
```
@ryanofsky 

Agree with all your suggestions.

As this PR has 3 ACKs already, is it acceptable to implement your suggestions in a following PR, e.g., #17453?

ryanofsky replied:
> As this PR has 3 ACKs already, is it acceptable to implement your suggestions in a following PR, e.g., #17453?

Of course yes, my suggestions are always just suggestions, and I wouldn't ACK a PR that I didn't think was ready to merge. I'd slightly prefer cleanup here or in a standalone PR instead of #17453 to make #17453 more focused and understandable, but it doesn't matter too much.

hebasto replied:
Done in the latest push.

hebasto committed:
refactor: Drop `bool force' parameter

hebasto committed:
qt: Rename SetPrune() to InitializePruneSetting()

This function now resets the prune size and is only called after the
intro sequence.

hebasto commented:
Implemented @ryanofsky's [suggestions](https://github.com/bitcoin/bitcoin/pull/17696#discussion_r363496670).

@Sjors @jonasschnelli @promag @ryanofsky 
Would you mind re-reviewing?

ryanofsky commented on src/qt/bitcoin.h:
```diff
@@ -67,8 +67,10 @@ class BitcoinApplication: public QApplication
-    void SetPrune(bool prune);
+    /// Initialize prune setting. If prune is set, intentionally override
+    /// existing prune size with the default size since this is called when
+    /// choosing a new datadir.
```
In commit "qt: Rename SetPrune() to InitializePruneSetting()" (192151ff193ceff8ad7177b07f0b98bf0ebcfc0d)

I think this comment would be a better placed inside the `InitializePruneSetting` method body because it's describing and justifying the implementation of the function, which (I think) is pretty confusing without a nearby explanation. I also think it would be better to add this comment in the earlier commit "qt: Force set nPruneSize in QSettings after intro" (6928c50fa3104ce1b77c5d12928d49584fa536d5) instead of here, because that is the commit which introduces the behavior this comment is describing.

hebasto replied:
Done.

ryanofsky commented on src/qt/optionsmodel.cpp:
```diff
@@ -252,14 +252,14 @@ void OptionsModel::SetPruneEnabled(bool prune, bool force)
 }
 
-void OptionsModel::SetPruneTargetGB(int prune_target_gb, bool force)
+void OptionsModel::SetPruneTargetGB(int prune_target_gb)
```
In commit "refactor: Drop `bool force' parameter" (0849b512916be3411db9d1efa447f2c12552ee0d)

Not a big deal, but I don't think the force argument should be dropped from this function. The call site in InitializePruneSetting is clearer if it has to explicitly set `force` to true, and it seems inconsistent and unsafe for one optionsmodel `SetPrune` method to soft-set by default while the other one force sets. The need to soft or force set depends on the context of the SetPrune call, and isn't something that seems right to hardcode here.

hebasto replied:
Done.

ryanofsky reviewed (approved):
Code review ACK 192151ff193ceff8ad7177b07f0b98bf0ebcfc0d. Just suggested changes since the last review and some minor related changes.

hebasto commented:
@ryanofsky 
Thank you for your review. All your comments have been addressed.

ryanofsky reviewed (approved):
Code review ACK af112ab62895b145660f4cd7ff842e9cfea2a530. Just suggested changes since last review (thanks!)

promag commented:
Tested ACK af112ab62895b145660f4cd7ff842e9cfea2a530. Latest suggestions and changes look good to me.

Once settings get stored in datadir I think the intro dialog could show prune target of the currently selected directory, if any (so it isn't overridden). Or disable/hide the prune option in the intro dialog if the currently selected directory exists.

Sjors commented:
Code review re-ACK af112ab62895b145660f4cd7ff842e9cfea2a530
